exec/vector/src/main/java/org/apache/drill/exec/vector/accessor/writer/AbstractTupleWriter.java

```diff
@@ -205,20 +208,29 @@ public int addColumnWriter(AbstractObjectWriter colWriter) {
 
   @Override
   public int addColumn(ColumnMetadata column) {
-    if (listener == null) {
-      throw new UnsupportedOperationException("addColumn");
-    }
+    verifyAddColumn(column.name());
     final AbstractObjectWriter colWriter = (AbstractObjectWriter) listener.addColumn(this, column);
     return addColumnWriter(colWriter);
   }
 
   @Override
   public int addColumn(MaterializedField field) {
+    verifyAddColumn(field.getName());
+    final AbstractObjectWriter colWriter = (AbstractObjectWriter) listener.addColumn(this, field);
```

Comment:
It would be nice if we could ged rid of casting to abstract version of object writer. We can add necessary methods to `ObjectWriter` interface or return `AbstractObjectWriter`. What do you think?

Reply:
Agreed this is awkward; but it is less awkward than the alternative. The idea is that the interfaces provide a clean interface for clients. I tried making the change, but then the implementation class, `AbstractObjectWriter` starts leaking through in many places. For example, the `getColumn()` method in the `TupleWriter` must return `AbstractObjectWriter`.

So, my thought was that the least intrusive solution is to simply cast to the implementation type the few places that it is needed rather than expose the implementation classes to clients.

Reply:
Well, since you used this approach before, I guess it ok to leave it as is but its definitely not Java-way, you kinda expect everybody who wants to implement `ObjectWriter` to extend `AbstractObjectWriter` instead and since this is not mandatory, this could lead to cast class exception in future.